Modules/_abc.c

```diff
@@ -441,154 +546,124 @@ _get_dump(PyObject *m, PyObject *args)
-static PyObject *
-_abc_init(PyObject *m, PyObject *args)
+// Compute set of abstract method names.
+static int
+compute_abstract_methods(PyObject *self)
 {
-    PyObject *ns, *bases, *keys, *abstracts, *base_abstracts, *self;
-    PyObject *key, *value, *item, *iter, *registry, *cache, *ref;
-    Py_ssize_t pos = 0;
-    int ret;
-    if (!PyArg_UnpackTuple(args, "_abc_init", 1, 1, &self)) {
-        return NULL;
-    }
-    /* Set up inheritance registry. */
-    ref = PyWeakref_NewRef(self, NULL);
-    if (!ref) {
-        return NULL;
-    }
-    registry = gset_new(&_guarded_set_type, NULL, NULL);
-    if (!registry) {
-        return NULL;
-    }
-    if (PyDict_SetItem(_the_registry, ref, registry) < 0) {
-        return NULL;
-    }
-    cache = PySet_New(NULL);
-    if (!cache) {
-        return NULL;
-    }
-    if (PyDict_SetItem(_the_cache, ref, cache) < 0) {
-        return NULL;
-    }
-    cache = PySet_New(NULL);
-    if (!cache) {
-        return NULL;
-    }
-    if (PyDict_SetItem(_the_negative_cache, ref, cache) < 0) {
-        return NULL;
-    }
-    if (PyDict_SetItem(_the_cache_version, ref, abc_invalidation_counter) < 0) {
-        return NULL;
-    }
-    if (!(abstracts = PyFrozenSet_New(NULL))) {
-        return NULL;
+    int ret = -1;
+    PyObject *abstracts = PyFrozenSet_New(NULL);
+    if (!abstracts) {
+        return -1;
     }
-    /* Compute set of abstract method names in two stages:
-       Stage 1: direct abstract methods.
+
+    PyObject *ns=NULL, *items=NULL, *bases=NULL;  // Py_CLEAR()ed on error.
+
+    /* Stage 1: direct abstract methods.
        (It is safe to assume everything is fine since type.__new__ succeeded.) */
     ns = _PyObject_GetAttrId(self, &PyId___dict__);
-    keys = PyMapping_Keys(ns); /* TODO: Fast path for exact dicts with PyDict_Next */
-    if (!keys) {
-        return NULL;
+    if (!ns) {
+        goto error;
+    }
+
+    /* TODO: Fast path for exact dicts with PyDict_Next */
+    items = PyMapping_Items(ns);
+    if (!items) {
+        goto error;
     }
-    for (pos = 0; pos < PySequence_Size(keys); pos++) {
-        key = PySequence_GetItem(keys, pos);
-        if (!key) {
-            Py_DECREF(keys);
+
+    for (Py_ssize_t pos = 0; pos < PyList_GET_SIZE(items); pos++) {
+        PyObject *it = PySequence_Fast(
+                PyList_GET_ITEM(items, pos),
+                "items() returned non-sequence item");
+        if (!it) {
             goto error;
         }
-        value = PyObject_GetItem(ns, key);
-        if (!value) {
-            Py_DECREF(keys);
-            Py_DECREF(key);
+        if (PySequence_Fast_GET_SIZE(it) != 2) {
+            PyErr_SetString(PyExc_TypeError, "items() returned not 2-tuple");
+            Py_DECREF(it);
             goto error;
         }
+
+        // borrowed
+        PyObject *key = PySequence_Fast_GET_ITEM(it, 0);
+        PyObject *value = PySequence_Fast_GET_ITEM(it, 1);
         int is_abstract = _PyObject_IsAbstract(value);
```

Comment:
Misuse of borrowed references; if value has a custom `__isabstractmethod__` property and items returns a list of lists (the elements of which will pass through `PySequence_Fast` unchanged, it the `_PyObject_IsAbstract` call could decref key or value out of existance. (and yes, for this to happen, the class passed would need to have a custom `__dict__` property that returns such a bad mapping)

Reply:
```
class K(ABCMeta):
    @property
    def __dict__(self):
        return BadItems()
class BadValue:
    @property
    def __isabstractmethod__(self):
        del ITEMS[0][:]
        return False
class BadItems(dict):
    def items(self):
        return ITEMS
ITEMS = [["x", BadValue()]]
```

Reply:
thanks.